File: Esame_232018/232018.c
/**
 * Realizzare un server (in C o in Java) che si metta in ascolto sul port 3333, sul quale accetta richieste di connessione da un client.
Sulla connessione stabilita, il server riceve una stringa, terminata dal carattere In° (si supponga che basti una sola operazione 
di lettura in ricezione da parte del server, per ricevere ciascuna stringa inviata dal client).
Il server quindi scrive la stringa ricevuta sullo standard output e, chiusa la connessione, si rimette in attesa di eventuali 
richieste di connessione.

Testare il server server_A usando telnet.


B. Conservando la versione (A) del server, realizzare un'ulteriore versione (B), 
tale che ciascuna stringa ricevuta dal server, oltre ad essere scritta dal server sulla propria standard output, 
sia inviata nuovamente al client come risposta (senza alcuna modifica).

Testare il server server_B usando telnet.
C. Conservando la versione (B) del server, realizzarne un'ulteriore versione tale che il server implementi un metodo/funzione
sommacifre (s)
che accetta per argomento una stringa se restituisce un intero; per il momento, 
ai fini di questo quesito (C), si supponga che, per qualunque stringa argomento s, venga restituito l'intero 0. 
In questa versione, il server, ricevuta una stringa s la scrive sulla standard output, calcola × = sommacifre (s) 
e invia al client un messaggio in forma di stringa del tipo n, x
dove n è un id progressivo della richiesta fatta dal client, mentre x è l'output del metodo sommaci fre (s). 
L'id, che fa parte dello stato del server, parte da 0 e viene incrementato ad ogni nuova richiesta da parte di un client 
fino a quando il server non viene riavviato.
Testare il server server_C usando telnet.
*/
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <netdb.h>
#include <ctype.h>


#define PORT 3333
#define MAX_SIZE_BUFFER 2048

int sommacifre(char *s){
    int sum = 0; 

    for(int i = 0; i<strlen(s); i++){
        if(isdigit(s[i])){
            sum += s[i] - '0'; 
        }
    }

    return sum ; 
}

int main(int argc, char **argv){
    char buffer[MAX_SIZE_BUFFER];
    struct sockaddr_in server, client; 
    socklen_t len = sizeof(client); 
    int s; 
    int check = 1; 
    int connectionSock;
   
    //Punto C
    int id = 0; 
    
    if((s = socket(AF_INET, SOCK_STREAM,0)) < 0){
        perror("Apertura socket\n");
        exit(-1); 
    }

    if((setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &check, sizeof(check))) < 0){
        perror("setsockopt()\n");
        exit(-1);
    }

    memset(&server, 0, sizeof(server)); 
    server.sin_family = AF_INET; 
    server.sin_addr.s_addr = INADDR_ANY; 
    server.sin_port = htons(PORT);


    if((bind(s, (struct sockaddr *)&server, sizeof(server))) < 0){
        perror("bind\n");
        exit(-1);
    }

    printf("Server ready.. \n"); 

    if(listen(s,1) != 0){
        perror("Listen.. \n"); 
        exit(-1);
    }

    if((connectionSock = accept(s, (struct sockaddr *)&client, &len)) < 0){
            perror("Connection accept()");
            exit(-1);
        }

    while(1){
        //A) 
        //accettiamo nuove connessioni con accept 
         int retcode;
        

        //leggiamo da connectionSock e lo inseriamo il buffer.
        if ( (retcode = read(connectionSock, &buffer, MAX_SIZE_BUFFER)) < 0 ) {
            perror("read()\n");
            exit(-1);
        }

        buffer[retcode] = '\0'; 
        printf("%s", buffer); 

        //Punto B fare echo 
        write(connectionSock, &buffer, strlen(buffer));

        //PUNTO C
        int sum = sommacifre(buffer); 

        id++; 

        snprintf(buffer, sizeof(buffer), "%d , %d\n\0", id, sum ); 

        write(connectionSock, &buffer, strlen(buffer)); 

    }

    close(connectionSock); 

}

File: Esame_232018/232018.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.ServerSocket;
import java.net.Socket;

public class Server {
    private static final int PORT = 3333; //costante
    private static final int MAX_SIZE_BUFFER = 2048; //costante 
    private static int requestCounter = 0; //id da incrementare

    public static void main(String[] args) {
        try {
            ServerSocket serverSocket = new ServerSocket(PORT);
            System.out.println("Server ready..");

            while (true) {
                /**Quando una connessione in ingresso viene ricevuta, viene creata un oggetto Socket denominato connectionSocket 
                 * per gestire questa connessione. 
                 * Il metodo accept() blocca l'esecuzione fino a quando non viene stabilita una connessione. */
                Socket connectionSocket = serverSocket.accept();
                /**Viene creato un BufferedReader per leggere i dati inviati dal client attraverso l'input stream del socket. */
                BufferedReader reader = new BufferedReader(new InputStreamReader(connectionSocket.getInputStream()));
                /**Viene creato un BufferedWriter per scrivere dati di risposta attraverso l'output stream del socket, che verranno inviati al client.*/
                BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(connectionSocket.getOutputStream()));

                String buffer;
                while ((buffer = reader.readLine()) != null) {
                    System.out.print(buffer+'\n');
                    // Calcoliamo la somma delle cifre nella stringa ricevuta
                    int sum = sommacifre(buffer);
                    //facciamo Eco
                    //writer.write(buffer + "\n"); 
                    //writer.flush();
                    // Inviamo la risposta al client
                    String response = (requestCounter++) + ", " + sum + "\n";
                    writer.write(response);
                    /**
                     * Questo comando forza il BufferedWriter a scrivere i dati effettivamente sullo stream del socket.
                     */
                    writer.flush();
                }

                connectionSocket.close();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Funzione per calcolare la somma delle cifre in una stringa
    private static int sommacifre(String s) {
        int sum = 0;
        for (int i = 0; i < s.length(); i++) {
            if (Character.isDigit(s.charAt(i))) {
                sum += Character.getNumericValue(s.charAt(i));
            }
        }
        return sum;
    }
}
